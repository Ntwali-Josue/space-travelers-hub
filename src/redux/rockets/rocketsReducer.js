import { createAsyncThunk } from '@reduxjs/toolkit';
import fetchAPI from '../fetchAPI';

const FETCH_ROCKETS = 'spaceX/rockets/FETCH_ROCKETS/fulfilled';
const RESERVE_ROCKET = 'spaceX/rockets/RESERVE_ROCKET';
const CANCEL_ROCKET = 'spaceX/rockets/CANCEL_ROCKET';

const initialState = {
  status: 'empty',
  rocketList: [],
};

export const fetchRockets = createAsyncThunk('spaceX/rockets/FETCH_ROCKETS', async () => {
  const response = await fetchAPI('https://api.spacexdata.com/v3/rockets');
  return response;
});

export const reserveRocket = (payload) => ({
  type: RESERVE_ROCKET,
  payload,
});

export const cancelRocket = (payload) => ({
  type: CANCEL_ROCKET,
  payload,
});

const rocketsReducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_ROCKETS:
      return {
        status: 'fetched',
        rocketList: [...state.rocketList, ...action.payload],
      };
    case RESERVE_ROCKET:
      return {
        status: state.status,
        rocketList: [...action.payload],
      };
    case CANCEL_ROCKET:
      return {
        status: state.status,
        rocketList: [...action.payload],
      };
    default:
      return state;
  }
};

export default rocketsReducer;
