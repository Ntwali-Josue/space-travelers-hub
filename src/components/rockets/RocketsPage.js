import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchRocket } from '../../redux/rockets/rocketsReducer';
import Rocket from './Rocket';

const RocketsPage = () => {
  const dispatch = useDispatch();
  const rockets = useSelector((state) => state.rockets.rocketList);
  const status = useSelector((state) => state.rockets.status);

  useEffect(() => {
    if (status === 'empty') {
      dispatch(fetchRocket());
    }
  }, [status]);

  const mapRockets = rockets.map(
    (rocket) => (
      <Rocket
        key={rocket.id}
        id={rocket.rocket_id}
        reserved={false}
        imgURL={`${rocket.flickr_images[1]}`}
        name={rocket.rocket_name}
        description={rocket.description}
      />
    ),
  );

  return (
    <section>
      {mapRockets}
      <p>.</p>
    </section>
  );
};

export default RocketsPage;
