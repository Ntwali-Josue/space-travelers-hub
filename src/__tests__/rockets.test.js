import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import rocketsReducer from '../redux/rockets/rocketsReducer';
import RocketsPage from '../components/rockets/RocketsPage';

describe('Rockets reducer', () => {
  test('initial state', () => {
    expect(rocketsReducer(undefined, {})).toEqual(
      {
        status: 'empty',
        rocketList: [],
        reservedRockets: [],
      }
    );
  });

  test('fetch rockets', () => {
    const initialState = {
      status: 'empty',
      rocketList: [],
      reservedRockets: [],
    };

    const action = {
      type: 'spaceX/rockets/FETCH_ROCKETS/fulfilled',
      payload: [{ name: 'Falcon 1'}, { name: 'Falcon 9'}],
    };

    expect(rocketsReducer(initialState, action)).toEqual(
      {
        status: 'fetched',
        rocketList: [{ name: 'Falcon 1'}, { name: 'Falcon 9'}],
        reservedRockets: [],
      }
    );
  });

  // test('initial state', () => {
  //   render(
  //     <Provider store={store}>
  //       <RocketsPage />
  //     </Provider>
  //   );
  //   expect(screen.getByText('section')).toMatchSnapshot();
  // });
});
