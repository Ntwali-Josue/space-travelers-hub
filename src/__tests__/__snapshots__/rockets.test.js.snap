// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets reducer App 1`] = `
<h1>
  Space Travelers Hub
</h1>
`;

exports[`Rockets reducer initial state 1`] = `
<h1>
  Space Travelers Hub
</h1>
`;

exports[`Rockets reducer initial state 2`] = `
<h2>
  Falcon 1
</h2>
`;

exports[`Rockets reducer render rockets 1`] = `
<h2>
  Falcon 1
</h2>
`;
